pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,932 @@
+		podIPv6CIDR:               nodeConfig.PodIPv6CIDR.String(),
+		nodeIPv4Addr:              nodeConfig.NodeIPv4Addr.IP.String(),
+		egressEnabled:             egressEnabled,
+		newBGPServerFn: func(globalConfig *bgp.GlobalConfig) bgp.Interface {
+			return gobgp.NewGoBGPServer(globalConfig)
+		},
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "bgpPolicy"),
+	}
+	c.bgpPolicyInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addBGPPolicy,
+			UpdateFunc: c.updateBGPPolicy,
+			DeleteFunc: c.deleteBGPPolicy,
+		},
+		resyncPeriod,
+	)
+	c.serviceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addService,
+			UpdateFunc: c.updateService,
+			DeleteFunc: c.deleteService,
+		},
+		resyncPeriod,
+	)
+	c.endpointSliceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addEndpointSlice,
+			UpdateFunc: c.updateEndpointSlice,
+			DeleteFunc: c.deleteEndpointSlice,
+		},
+		resyncPeriod,
+	)
+	if c.egressEnabled {
+		c.egressInformer = egressInformer.Informer()
+		c.egressLister = egressInformer.Lister()
+		c.egressListerSynced = egressInformer.Informer().HasSynced
+		c.egressInformer.AddEventHandlerWithResyncPeriod(
+			cache.ResourceEventHandlerFuncs{
+				AddFunc:    c.addEgress,
+				UpdateFunc: c.updateEgress,
+				DeleteFunc: c.deleteEgress,
+			},
+			resyncPeriod,
+		)
+	}
+	c.nodeInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addNode,
+			UpdateFunc: c.updateNode,
+			DeleteFunc: nil,
+		},
+		resyncPeriod,
+	)
+
+	c.secretInformer = coreinformers.NewFilteredSecretInformer(k8sClient,
+		env.GetAntreaNamespace(),
+		resyncPeriod,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		func(options *metav1.ListOptions) {
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", types.BGPPolicySecretName).String()
+		})
+	c.secretInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addSecret,
+		UpdateFunc: c.updateSecret,
+		DeleteFunc: c.deleteSecret,
+	})
+
+	return c, nil
+}
+
+func (c *Controller) Run(ctx context.Context) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting", "controllerName", controllerName)
+	defer klog.InfoS("Shutting down", "controllerName", controllerName)
+
+	cacheSyncs := []cache.InformerSynced{
+		c.nodeListerSynced,
+		c.serviceListerSynced,
+		c.bgpPolicyListerSynced,
+		c.endpointSliceListerSynced,
+		c.serviceListerSynced,
+	}
+	if c.egressEnabled {
+		cacheSyncs = append(cacheSyncs, c.egressListerSynced)
+	}
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), cacheSyncs...) {
+		return
+	}
+
+	go wait.UntilWithContext(ctx, c.worker, time.Second)
+
+	go c.secretInformer.Run(ctx.Done())
+	cache.WaitForCacheSync(ctx.Done(), c.secretInformer.HasSynced)
```

Comment:
Shouldn't it be waited before starting the worker, along with other informers? otherwise the worker would try to connect peer routers without providing any passwords first.

Reply:
Make sense. Moved these two lines before `go wait.UntilWithContext(ctx, c.worker, time.Second)`